src/Coordination/Changelog.cpp

```diff
@@ -585,131 +657,844 @@ class ChangelogReader
+            break;
+        else if (changelog_description == current_file_info->file_description)
+            ++current_file_info->count;
+        else
+            current_file_info = &file_infos.emplace_back(changelog_description, position, /* count */ 1);
+
+        total_size += size;
+        commit_logs_cache.addPrefetchedEntry(current_index, size);
+    }
+
+    if (!file_infos.empty())
+    {
+        current_prefetch_info = std::move(new_prefetch_info);
+        prefetch_to = current_index - 1;
+        LOG_TRACE(log, "Will prefetch {} commit log entries [{} - {}]", prefetch_to - prefetch_from + 1, prefetch_from, prefetch_to);
+
+        current_prefetch_info->file_infos = std::move(file_infos);
+        auto inserted = prefetch_queue.push(current_prefetch_info);
+        chassert(inserted);
+    }
+}
+
+CacheEntry::CacheEntry(LogEntryPtr entry_)
+    : entry(std::move(entry_))
+{
+    if (entry == nullptr)
+        is_prefetched = true;
+}
 
-        LOG_TRACE(log, "Reading from disk {}", disk->getName());
-        std::unordered_map<std::string, std::string> incomplete_files;
+LogEntryStorage::InMemoryCache::InMemoryCache(size_t size_threshold_)
+    : size_threshold(size_threshold_)
+{}
+
+void LogEntryStorage::InMemoryCache::updateStatsWithNewEntry(uint64_t index, size_t size)
+{
+    cache_size += size;
+
+    if (cache.size() == 1)
+    {
+        min_index_in_cache = index;
+        max_index_in_cache = index;
+    }
+    else
+    {
+        max_index_in_cache = index;
+    }
+}
 
-        const auto clean_incomplete_file = [&](const auto & file_path)
+void LogEntryStorage::InMemoryCache::addEntry(uint64_t index, LogEntryPtr log_entry)
+{
+    auto entry_size = logEntrySize(*log_entry);
+    auto [_, inserted] = cache.emplace(index, std::move(log_entry));
+    if (!inserted)
+    {
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Trying to insert log with index {} which is already present in cache", index);
+    }
+    updateStatsWithNewEntry(index, entry_size);
+}
+
+void LogEntryStorage::InMemoryCache::addEntry(IndexToCacheEntryNode && node)
+{
+    auto index = node.key();
+    auto entry_size = logEntrySize(*node.mapped().entry);
+
+    auto result = cache.insert(std::move(node));
+    if (!result.inserted)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Trying to insert log with index {} which is already present in cache", index);
+
+    updateStatsWithNewEntry(index, entry_size);
+}
+
+void LogEntryStorage::InMemoryCache::addPrefetchedEntry(uint64_t index, size_t size)
+{
+    auto [_, inserted] = cache.emplace(index, nullptr);
+    if (!inserted)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Trying to set prefetched entry with index {} which is already present in cache", index);
+    updateStatsWithNewEntry(index, size);
+}
+
+void LogEntryStorage::InMemoryCache::setPrefetchedEntry(uint64_t index, LogEntryPtr log_entry, std::exception_ptr exception)
```

Comment:
Rename and I would prefer different methods for exceptions and normal entries.

Reply:
At least assert in current implementation like `chassert(log_entry != nullptr || exception != nullptr)`